docs/source/json-api/lf-value-specification.rst

```diff
@@ -418,13 +418,13 @@ and then, from JSON
 
 ::
 
-    {"Bar": {"f1": 42, "f2": true}}
-    {"Baz": {}}
+    {"tag": "Bar", "value": {"f1": 42, "f2": true}}
+    {"tag": "Baz", "value": {}}
 
 This can be encoded and used in TypeScript, including exhaustiveness
-checking; see `a keyed example`_.
+checking; see `a type refinement example`_.
 
-.. _a keyed example: https://www.typescriptlang.org/play/#src=type%20Foo%20%3D%0D%0A%20%20%20%20%7B%20Bar%3A%20%7B%20f1%3A%20number%2C%20f2%3A%20boolean%20%7D%20%7D%0D%0A%20%20%7C%20%7B%20Baz%3A%20%7B%20f3%3A%20string%20%7D%20%7D%3B%0D%0A%0D%0Afunction%20test(v%3A%20Foo)%20%7B%0D%0A%20%20if%20(%22Bar%22%20in%20v)%20%7B%0D%0A%20%20%20%20console.log(v.Bar.f1%2C%20v.Bar.f2)%3B%0D%0A%20%20%7D%20else%20if%20(%22Baz%22%20in%20v)%20%7B%0D%0A%20%20%20%20console.log(v.Baz.f3)%3B%0D%0A%20%20%7D%20else%20%7B%0D%0A%20%20%20%20const%20_%3A%20never%20%3D%20v%3B%0D%0A%20%20%7D%0D%0A%7D%20%0D%0A
+.. _a type refinement example: https://www.typescriptlang.org/play/?ssl=10&ssc=4&pln=8&pc=5#code/C4TwDgpgBAYg9nKBeAsAKCpqBvKwCGA5gFxQBEAQvgE5kA0UAbvgDYCuEpuAZgIykA7NgFsARhGoNuAJlKiELCPgFQAvmvSYAPjjxFSlfAC96TVhy5q1AbnTpubAQGNgASzgrgEAM7AAFIyk8HAAlDiaUK7cUAEAdASEyEhI5FS0YdgRmE4e3nCKsSxwhHHM7BCxfAyMsWUcldIhthhWECze0FExNQlJKYYmGVlQOQJ5BUUlNXUQTRHqbR3hLdm5wFAA+oIQjBLITM2Yqujq6EA
```

Comment:
Maybe using `switch`/`case`:

```suggestion
.. _a type refinement example: https://www.typescriptlang.org/play/#code/C4TwDgpgBAYg9nKBeAsAKCpqBvKwCGA5gFxQBEAQvgE5kA0UAbvgDYCuEpuAZgIykA7NgFsARhGoNuAJlKiELCPgFQAvmvSYAPjjxFSlfAC96TVhy5q1AbnTpubAQGNgASzgrgEAM7AAFIyk8HAAlDiaUN4A7q7ATgAWUAEAdASEYdgRmE743tCGtMRZWE4e3nCKySxwhCnM7BDJfAyMyfUcTdIhthhYmNQQwGzUAj19OXnkVCZFveNlFY3Vta3tEN3F-YPDo8UAJhDc+GwswLN92WXAUAD6ghCMEshMYxpoqkA
```

Reply:
There's no need for the `default` branch. TypeScripts pattern match exhaustiveness checker will yell at you even if you omit it.

Reply:
@hurryabit I don't observe such an error or warning [for this example](https://www.typescriptlang.org/play/#code/C4TwDgpgBAYg9nKBeAsAKCpqBvKwCGA5gFxQBEAQvgE5kA0UAbvgDYCuEpuAZgIykA7NgFsARhGoNuAJlKiELCPgFQAvmvSYAPjjxFSlfAC96TVhy5q1AbnTpubAQGNgASzgrgEAM7AAFIyk8HAAlDiaUN4A7q7ATgAWUAEAdASEYdgRmE743tCGtMRZWE4e3nCKySxwhCnM7BDJfAyMyfUcTdIhthhYmNQQwGzUAj19APQAVDl55FQmRb19pQLlldW1re0Q3cX9g8Ojk+MRquiqQA). Do you have to pass a special typechecker flag?

Reply:
Interesting. The exhaustiveness checker does not fire for functions "returning" `void`. If you return a proper value, say `boolean`, you'll get an error. See [this example](https://www.typescriptlang.org/play/?ssl=9&ssc=1&pln=11&pc=1#code/C4TwDgpgBAYg9nKBeAsAKCpqBvKwCGA5gFxQBEAQvgE5kA0UAbvgDYCuEpuAZgIykA7NgFsARhGoNuAJlKiELCPgFQAvmvSYAPjjxFSlfAC96TVhy5q1AbnTpubAQGNgASzgrgEAM7AAFIyk8HAAlDiaUN4A7q7ATgAWUAEAdASEYdgRmE743tCGtMRZWNQQwGzUKozJzOwQyTK2GFgA9C1QOXnkVCZFzZhtJWUVKtyseU2YquiqQA).

Reply:
Let's keep the `default` branch then.

Reply:
It seems the `switch` body itself must also be wholly responsible for the return value, [as in this example](https://www.typescriptlang.org/play/#code/C4TwDgpgBAYg9nKBeAsAKCpqBvKwCGA5gFxQBEAQvgE5kA0UAbvgDYCuEpuAZgIykA7NgFsARhGoNuAJlKiELCPgFQAvmvSYAPjjxFSlfAC96TVhy5q1AbnTpubAQGNgASzgrgEAM7AAFIyk8HAAlDiaUN4A7q7ATgAWUAEAdASEYdgRmE743tCGtMRZWNQQwGzUKozJzOwQyTK2GFgA9C1QOXnkVCZFzZhtJWUVKtyseU2YqhGl5ZVQYywT6KpAA).

I don't know what Typescript common practice is, but I think for my own style this would imply using the no-`default` shortcut wherever possible (which naturally depends on a functional style) but including the `default` `v: never` case otherwise. For this example I think presenting only the latter is fine (we are not writing a Typescript tutorial after all) but we could include both.